crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//!
+//! ### Rolling aggregation
+//!
+//! By using a "moving average" to average recent data,
+//! we can obtain a dataset with less noise due to variation from month
+//! to month.   DBSP
+//! provides [`Stream::partitioned_rolling_average`] for this purpose.  To
+//! use it, we have to map our indexed Z-set into a "partitioned indexed
+//! Z-set" ([`PartitionedIndexedZSet`]).
+//! This is just an indexed Z-set in which the key is the "partition"
+//! within which averaging occurs (for us, this is the country), and the
+//! value is a tuple of a "timestamp" and a value.  DBSP uses the
+//! timestamp component, which must have an unsigned integer type, to
+//! define the window.  We can map our `(country, year, month)` key and
+//! `vaccinations` value into a `country` partition, `date` timestamp, and
+//! `vaccinations` value like this:
+//!
+//! ```ignore
+//!     let moving_averages = monthly_totals
+//!         .map_index(|((c, y, m), v)| (c.clone(), (*y as u32 * 12 + (*m as u32 - 1), *v)))
+//! ```
+//!
+//! Once we've done that, computing the moving average is easy.  Here's how we
+//! get the average of the current month and the two preceding months (when
+//! they're in the data set):
+//!
+//! ```ignore
+//!         .partitioned_rolling_average(RelRange::new(RelOffset::Before(2), RelOffset::Before(0)))
+//! ```
+//!
+//! We can re-map to recover year and month from the timestamp that we made.  We
+//! also strip off a `Some(...)` that gets wrapped around the rolling average.
+//! This can end up as a `None` if there are no rows in the window, but, in
+//! turn, that can only happen if the window doesn't include the current row,
+//! and we didn't define our window that way.  Anyhow:
+//!
+//! ```ignore
+//!         .map_index(|(c, (date, avg))| ((c.clone(), date / 12, date % 12 + 1), avg.unwrap()));
+//! ```
+//!
+//! We can print the data the same way as before:
+//!
+//! ```ignore
+//! moving_averages.inspect(|data| {
+//!     data.iter()
+//!         .for_each(|((c, y, m), sum, w)| println!("{c:16} {y}-{m:02} {sum:10}: {w:+}"))
+//! });
+//! ```
+//!
+//! The output looks like this (you can verify that the second row is the
+//! average of the first two rows in the previous output, and so on):
+//!
+//! ```text
+//! England          2021-01 5600174: +1
+//! England          2021-02 7488796: +1
+//! England          2021-03 8946255: +1
+//! England          2021-04 10842512: +1
+//! England          2021-05 12307688: +1
+//! England          2021-06 12002268: +1
+//! ...
+//! Northern Ireland 2021-01 150315: +1
+//! Northern Ireland 2021-02 233694: +1
+//! ...
+//! Wales            2021-01 295057: +1
+//! Wales            2021-02 458273: +1
+//! Wales            2021-03 584463: +1
+//! ```
+//!
+//! The whole program is in `tutorial6.rs`.
+//!
+//! ### Joins
+//!
+//! Suppose we want both the current month's vaccination count and the moving
+//! average together. With enough work, we could get them in a linear series of
```

Comment:
"linear series of steps" makes you think about the linear aggregate above.
I would rather remove this comment altogether and say that we do it with a join.

Reply:
I changed this to:

```
//! Suppose we want both the current month's vaccination count and the moving
//! average together. With enough work, we could get them with just
//! aggregation by writing our own "aggregator" ([`Aggregator`]).  It's a
//! little easier to do a join, and it gives us a chance to show how to do that.
```